keyboards/model01/rules.mk

```diff
@@ -0,0 +1,65 @@
+SRC += i2c_master.c
+
+# MCU name
+MCU = atmega32u4
+
+# Processor frequency.
+#     This will define a symbol, F_CPU, in all source code files equal to the
+#     processor frequency in Hz. You can then use this symbol in your source code to
+#     calculate timings. Do NOT tack on a 'UL' at the end, this will be done
+#     automatically to create a 32-bit value in your source code.
+#
+#     This will be an integer division of F_USB below, as it is sourced by
+#     F_USB after it has run through any CPU prescalers. Note that this value
+#     does not *change* the processor frequency - it should merely be updated to
+#     reflect the processor speed set externally so that the code can use accurate
+#     software delays.
+F_CPU = 16000000
+
+
+#
+# LUFA specific
+#
+# Target architecture (see library "Board Types" documentation).
+ARCH = AVR8
+
+# Input clock frequency.
+#     This will define a symbol, F_USB, in all source code files equal to the
+#     input clock frequency (before any prescaling is performed) in Hz. This value may
+#     differ from F_CPU if prescaling is used on the latter, and is required as the
+#     raw input clock is fed directly to the PLL sections of the AVR for high speed
+#     clock generation for the USB and other AVR subsections. Do NOT tack on a 'UL'
+#     at the end, this will be done automatically to create a 32-bit value in your
+#     source code.
+#
+#     If no clock division is performed on the input clock inside the AVR (via the
+#     CPU clock adjust registers or the clock division fuses), this will be equal to F_CPU.
+F_USB = $(F_CPU)
+
+# Interrupt driven control endpoint task(+60)
+OPT_DEFS += -DINTERRUPT_CONTROL_ENDPOINT
+
+BOOTLOADER = caterina
+
+# Build Options
+#   change yes to no to disable
+#
+BOOTMAGIC_ENABLE = no      # Virtual DIP switch configuration(+1000)
+MOUSEKEY_ENABLE = no       # Mouse keys(+4700)
+EXTRAKEY_ENABLE = yes       # Audio control and System control(+450)
+CONSOLE_ENABLE = yes        # Console for debug(+400)
+COMMAND_ENABLE = no        # Commands for debug and configuration
+# Do not enable SLEEP_LED_ENABLE. it uses the same timer as BACKLIGHT_ENABLE
+SLEEP_LED_ENABLE = no       # Breathing sleep LED during USB suspend
+# if this doesn't work, see here: https://github.com/tmk/tmk_keyboard/wiki/FAQ#nkro-doesnt-work
+NKRO_ENABLE = no            # USB Nkey Rollover
+BACKLIGHT_ENABLE = no       # Enable keyboard backlight functionality on B7 by default
+RGBLIGHT_ENABLE = no        # Enable keyboard RGB underglow
+MIDI_ENABLE = no            # MIDI support (+2400 to 4200, depending on config)
+UNICODE_ENABLE = no         # Unicode
+BLUETOOTH_ENABLE = no       # Enable Bluetooth with the Adafruit EZ-Key HID
+AUDIO_ENABLE = no           # Audio output on port C6
+FAUXCLICKY_ENABLE = no      # Use buzzer to emulate clicky switches
+HD44780_ENABLE = no 		# Enable support for HD44780 based LCDs (+400)
+
+CUSTOM_MATRIX = yes
```

Comment:
This specifies a custom matrix, but never actually includes it.

Reply:
The matrix is defined in `model01.c`, so it is included without further ado. Would you prefer a separate file? I originally decided not to split files based on the small total size of the code.

Reply:
Honestly, that's up to you.  

But most (if not all) of the boards that use a custom matrix have a `matrix.c` file specifically for the matrix, and then the keyboard's c file for ... more keyboard related stuff.

Reply:
I think that probably makes sense just because someone will probably come and want to implement a bunch of LED features some day... latest commit splits out LED and matrix, leaving one line of actual code which belongs in `model01.c` ;-)

Reply:
Yup, exactly.  Like myself. 🤣 
And the way it's set up there, it should be simple to implement something with the matrix scan function, per keymap. 

Though, a good question is how bad is the performance hit for each `set_led_to_raw` command? 
Eg, if you call it 50 times in one matrix scan, will it be noticeable?

Reply:
Actually, I don't see any noticeable latency scanning out all 64 LEDs this way! Impressive.
There are block transfer functions available that increase the efficiency. But I guess that's not really necessary!

Reply:
Ah, my mistake: they need a delay between transactions. I was posting all the writes but they weren't all taking effect ^_^

Reply:
It looks like it works okay with a 10µs delay after each write. I'll plonk that in the write functions.